pallets/rmrk-core/src/lib.rs

```diff
@@ -395,6 +399,21 @@ pub mod pallet {
 			}
 			.unwrap_or_default();
 
+			let (root_owner, root_nft) = Pallet::<T>::lookup_root_owner(collection_id, nft_id)?;
+			// Check ownership
+			ensure!(sender == root_owner, Error::<T>::NoPermission);
+			// Prepare transfer
+			let new_owner_account = match new_owner.clone() {
+				AccountIdOrCollectionNftTuple::AccountId(id) => id,
+				AccountIdOrCollectionNftTuple::CollectionAndNftTuple(cid, nid) => {
+					// Circle detection
+					let (_, new_owner_root_nft) = Pallet::<T>::lookup_root_owner(cid, nid)?;
+					ensure!(root_nft != new_owner_root_nft, Error::<T>::CircleDetected);
+					// Convert to virtual account
+					Pallet::<T>::nft_to_account_id::<T::AccountId>(cid, nid)
+				},
+			};
```

Comment:
With the lazy ownership, we don't need to recursively change the owner of the NFT anymore. The logic in `nft_send` should be changed